### ISSUE TITLE
rootcling compilation failure during standalone build (low priority)

### ISSUE THREAD

**Issue by srlantz** [trackreco/cmssw#117]
Something isn't quite working about `rootcling`'s include paths during standalone build. The easy workaround is just to leave `WITH_ROOT` undefined for `make`.

```
...
make[1]: Entering directory '/home/users/slantz/mkFit-standalone/objs-CMS'
rootcling -f ShellDict.cc /home/users/slantz/mkFit-standalone/RecoTracker/MkFitCMS/standalone/Shell.h /home/users/slantz/mkFit-standalone/RecoTracker/MkFitCMS/standalone/ShellLinkDef.h
In file included from input_line_9:3:
/home/users/slantz/mkFit-standalone/RecoTracker/MkFitCMS/standalone/Shell.h:4:10: fatal error: 'RecoTracker/MkFitCore/interface/Hit.h' file not found
#include "RecoTracker/MkFitCore/interface/Hit.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error: rootcling: compilation failure (./ShellDict8148ac5312_dictUmbrella.h)
make[1]: *** [/home/users/slantz/mkFit-standalone/RecoTracker/MkFitCMS/standalone/Makefile:31: ShellDict.cc] Error 1
make[1]: Leaving directory '/home/users/slantz/mkFit-standalone/objs-CMS'
make: *** [Makefile:15: all] Error 2
```

The supposedly missing include file really does exist at the given location. The mkFit code version I'm using is from a recent head of CMSSW, commit cbe3152 on 2/21/23. ROOT is obtained from cvmfs by `source`-ing the following script (the `if` condition is slightly modified for cgpu-1, based on the current RecoTracker/MkFitCore/standalone/xeon_scripts/init-env.sh):

```
[1539] slantz@cgpu-1 ~/mkFit-standalone$ cat RecoTracker/MkFitCore/standalone/xeon_scripts/init-env.sh
#!/bin/bash
if [[ `[ -f /usr/bin/lsb_release ] && lsb_release -si` == \
        "Fedora" ||
      `[ -f /etc/redhat-release ] && awk '{print $1}' /etc/redhat-release` == \
        "AlmaLinux" ]]
then
  source /cvmfs/cms.cern.ch/el8_amd64_gcc10/lcg/root/6.24.07-da610b2b7ed663a0a05d3605f3d83ceb/etc/profile.d/init.sh
  export TBB_GCC=/cvmfs/cms.cern.ch/el8_amd64_gcc10/external/tbb/v2021.5.0-e966a5acb1e4d5fd7605074bafbb079c/
else
  source /cvmfs/cms.cern.ch/slc7_amd64_gcc10/lcg/root/6.24.07-f52350f4e0b802edeb9a2551a7d00b92/etc/profile.d/init.sh
  export TBB_GCC=/cvmfs/cms.cern.ch/slc7_amd64_gcc10/external/tbb/v2021.4.0-d0152ca29055e3a1bbf629673f6e97c4
fi
```
The resulting rootcling isn't causing the problem, I think, but I don't know much about it:
```
[1541] slantz@cgpu-1 ~/mkFit-standalone$ which rootcling
/cvmfs/cms.cern.ch/el8_amd64_gcc10/lcg/root/6.24.07-da610b2b7ed663a0a05d3605f3d83ceb/bin/rootcling
```
Otherwise I am following the instructions in RecoTracker/MkFitCore/standalone/README_buildFromCMSSW.txt, which I believe should still be valid (given the above modification to `init-env.sh`), and which seem to work when `WITH_ROOT` is left undefined.

**Comment by osschar** [comment]
I used setup from one of the latest cmssw pre-13 releases:
```
SCRAM_ARCH=el8_amd64_gcc11
ROOTSYS=/cvmfs/cms.cern.ch/el8_amd64_gcc11/lcg/root/6.26.07-3c8fea8a0ce2aca35570ac22afed05d0/
```
Apparently the newer rootcling does not check the input header file but just places them in the generated dictionary.cc file.

I don't know if we should fix this or just use the newer ROOT, I added `-I${SRCDIR}` to rootcling lines in `MkFitCMS/standalone/Makefile` and it produced the same thing ... probably it would fix the compilation for you.

As I was trying this, I noticed another change got missed in the above Makefile, namely `DictsDict.o` should have been changed to `WriteMemFileDict.o`. Below is the full diff.

```
diff --git a/RecoTracker/MkFitCMS/standalone/Makefile b/RecoTracker/MkFitCMS/standalone/Makefile
index e385aaeee3a..4b3e20b2214 100644
--- a/RecoTracker/MkFitCMS/standalone/Makefile
+++ b/RecoTracker/MkFitCMS/standalone/Makefile
@@ -23,12 +23,12 @@ SRCS := $(wildcard ${CMS_DIR}/src/*.cc) $(wildcard ${SACMS}/*.cc)
 ifdef WITH_ROOT
 SRCS += ${SACMS}/tkNtuple/WriteMemoryFile.cc
 WriteMemFileDict.cc ${WMF_DICT_PCM}: ${SACMS}/tkNtuple/DictsLinkDef.h
-       rootcling -f WriteMemFileDict.cc $<
+       rootcling -I${SRCDIR} -f WriteMemFileDict.cc $<
        mv WriteMemFileDict_rdict.pcm ${WMF_DICT_PCM}

 SRCS += ShellDict.cc
 ShellDict.cc ${SHELL_DICT_PCM}: ${SACMS}/Shell.h ${SACMS}/ShellLinkDef.h
-       rootcling -f ShellDict.cc ${SACMS}/Shell.h ${SACMS}/ShellLinkDef.h
+       rootcling -I${SRCDIR} -f ShellDict.cc ${SACMS}/Shell.h ${SACMS}/ShellLinkDef.h
        mv ShellDict_rdict.pcm ${SHELL_DICT_PCM}
 endif
 SRCB := $(notdir ${SRCS})
@@ -61,7 +61,7 @@ ${LIB_CMS}: ${CMS_OBJS}
 ${MAIN}: ${LIB_CMS} mkFit.o
        ${CXX} ${CXXFLAGS} ${VEC_HOST} ${LDFLAGS} mkFit.o -o $@ ${LDFLAGS_HOST} -ltbb -L.. -lMicCore -lMicCMS -Wl,-rpath=.

-${WRMEMF}: WriteMemoryFile.o DictsDict.o
+${WRMEMF}: WriteMemoryFile.o WriteMemFileDict.o
        ${CXX} ${CXXFLAGS} ${LDFLAGS} $^ -o $@ ${LDFLAGS_HOST} -ltbb -L.. -lMicCore -Wl,-rpath=.

 ${OBJS}: %.o: %.cc %.d
```

**Comment by srlantz** [comment]
@osschar - Implementing your diff of the Makefile worked. I modified your option to `-I=${SRCDIR}`, since the `-I=` syntax is what is documented in `rootcling --help`. (Your way without the `=` works too.) Also, I edited `init-env.sh` to grab ROOT (and TBB) from the `el8_amd64_gcc11` tree under cms.cern.ch as you recommended, instead of `el8_amd64_gcc10` like we’ve been using:
```
  source /cvmfs/cms.cern.ch/el8_amd64_gcc11/lcg/root/6.26.07-3c8fea8a0ce2aca35570ac22afed05d0/etc/profile.d/init.sh
  export TBB_GCC=/cvmfs/cms.cern.ch/el8_amd64_gcc10/external/tbb/v2021.5.0-36aff7df349e0716374b1668ccd18e17/
```
Just by itself, the newer ROOT 6.26 doesn't seem to do the trick. It’s the `-I=${SRCDIR}` option that appears to be necessary. Without it, compilation fails with the same error. As a matter of fact, if the Makefile is edited as you indicated, ROOT 6.24 and gcc-10 seem to work just fine for compiling too.

Thank you for providing `/home/scratch/ttbar-pu50-1000.bin` on cgpu-1. The current mkFit expects file version 7 as input, and your file works fine. With that as input, I find that mkFit runs 5-7% slower than a mid-November commit, where gcc-11 is producing the slightly better number. Also, speedup with `-Ofast` is not as good as the earlier code: it’s only ~15% faster instead of ~20%. (The test is with 1 thread, 100 events, AVX2.)

Finally, as I mentioned, the standalone `init-env.sh` script currently grabs ROOT and TBB from the `el8_amd64_gcc10` tree. Are you implying that we'll want to move ahead with gcc11 for compiling, given that this is the environment you prefer to set for pre-13? I don't know what is expected for CMSSW these days.

If we stick with gcc10, and we try to move up from ROOT 6.24 to 6.26, then only one build of ROOT is available, and its hash doesn't match the one you proposed:
```
$ ls -l /cvmfs/cms.cern.ch/el8_amd64_gcc10/lcg/root
[...]
drwxr-xr-x 18 cvmfs cvmfs 224 Aug 28 05:05 6.26.07-3ba3a89d151664634e49fcc4c2c977f5
```
I don't know... Maybe it makes most sense to use gcc-11 and ROOT 6.26 for pre-13, and gcc-10 and ROOT 6.24 for the current head of CMSSW?
